students/Briadean/lesson09/OOMailroom.py

```diff
@@ -0,0 +1,177 @@
+            print(donor)
+
+    def generate_report(self):
+        """Build a report of donors"""
+        report_lines = []
+        for donor in self._donors.values():
+            total_given = sum(donor.donations)
+            num_gifts = len(donor.donations)
+            average_gifts = total_given / num_gifts
+            report_lines.append((donor.name, total_given, num_gifts, average_gifts))
+
+        report_header = "\n{:20} | {:12} | {:11} | {:12}".format("Donor Name",
+                                                                 "Total Given",
+                                                                 "Num Gifts",
+                                                                 "Average Gifts"
+                                                                 )
+
+        report_lines.sort(key=self._sort_key, reverse=True)
+        report = []
+        report.append(report_header)
+        report.append("-" * 55)
+        [report.append("{:20}  ${:12,.2f}  {:11}  ${:12,.2f}".format(*line)) for line in report_lines]
+
+        print("\n".join(report))
+
+    @staticmethod
+    def _sort_key(item):
+        """Sort function
+        Params: input = "item" or "entry" in the dictionary
+                returns the "Total Given" column entry
+        """
+        return item[1]
+
+    def generate_thank_you(self, name, donation):
+        """Generates a standardized thank you letter
+        Params: input = donor name, most recent donation amounts
+                returns a write_letters_disk
+        """
+        letter = ("\nDear {},\n\n"
+                  "Thank you for your recent generous donation of ${:,.2f}.\n"
+                  "It will be put to good use.\n\n"
+                  "Sincerely,\n"
+                  "The Resistance Against Giant Flying Meatballs\n".format(name, donation))
+        return letter
+
+    def write_letters_disk(self):
+        """Writes a thank you letter for each donor in the database to disk"""
+        for donor, donation in donors_dict.items():
+            letter = self.generate_thank_you(donor, donation[-1])
+            filename = donor + ".txt"
+            print("Writing letter for:", donor)
+            open(filename, "w").write(letter)
+
+    def accept_donation(self, donor_name):
+        """Accept a new donation.
+        Params: input = donor name
+        """
+        donation_amt = float(input("\nHow much are they donating? >>$ "))
+        try:
+            self._donors[donor_name].add_donation(donation_amt)
+        except KeyError:
+            donor = Donor(name=donor_name, donations=[donation_amt])
+            self.add_donor(donor)
+
+
+def seed_db(db):
+    for name, donation in donors_dict.items():
+        donor = Donor(name=name, donations=donation)
+        db.add_donor(donor)
+
+
+def main():
+    db = DonorDB()
+    seed_db(db)
+
+    menu_dict = {1: thank_you_menu,
+                 2: db.write_letters_disk,
+                 3: db.generate_report,
```

Comment:
Interesting usage of methods as function names, I'd say, as I myself have never seen this before, but it looks making sense. Maybe this is going to create a curried function, not requiring the self param anyway. Learned something new from this. Thanks.